django_lti_tool_provider/views.py

```diff
@@ -0,0 +1,148 @@
+
+    def get(self, request, *args, **kwargs):
+        return self.process_request(request)
+
+    def post(self, request, *args, **kwargs):
+        return self.process_request(request)
+
+    def process_request(self, request):
+        if not request.user.is_authenticated():
+            try:
+                lti_parameters = self._get_lti_parameters_from_request(request)
+                lti_data = {
+                    hook_name: lti_parameters.get(lti_name, None)
+                    for lti_name, hook_name in self.PASS_TO_AUTHENTICATION_HOOK.items()
+                }
+            except oauth2.Error, e:
+                _logger.exception(u"Invalid LTI Request")
+                return HttpResponseBadRequest("Invalid LTI Request: " + e.message)
+
+            for hook in self.authentication_hooks:
+                hook(request, lti_data)
+
+        if request.user.is_authenticated():
+            return self.process_authenticated_lti(request)
+        else:
+            return self.process_anonymous_lti(request)
+
+    @classmethod
+    def _get_lti_parameters_from_request(cls, request):
+        provider = DjangoToolProvider(settings.CONSUMER_KEY, settings.LTI_SECRET, request.POST)
+        provider.valid_request(request)
+        return provider.to_params()
+
+    @classmethod
+    def _get_anonymous_redirect_url(cls, request):
+        try:
+            result = "{login}?next={lti_handler}".format(
+                login=reverse('django.contrib.auth.views.login'),
+                lti_handler=request.path.strip('/')
+            )
+        except NoReverseMatch:
+            result = "/"
+        return result
+
+    @classmethod
+    def _get_authenticated_redirect_url(cls):
+        redirect_to = getattr(settings, 'REDIRECT_AFTER_LTI', '')
+        try:
+            result = reverse(redirect_to)
+        except NoReverseMatch:
+            result = "/"+redirect_to
+        return result
+
+    @classmethod
+    def register_authentication_hook(cls, hook):
+        """ Adds hook to pre-processing hooks list """
+        cls.authentication_hooks.append(hook)
+
+    @classmethod
+    def process_anonymous_lti(cls, request):
+        """
+        This method handles LTI request if it was sent prior to MyDante authorization. In such a case, we need user
+        authenticated first. Unfortunately, it looses POST data in the process, so when it gets back original LTI
+        request is gone. So we save important parts of it into session to retrieve when OpenID authentication happened
+        """
+        try:
+            provider = DjangoToolProvider(settings.CONSUMER_KEY, settings.LTI_SECRET, request.POST)
+            provider.valid_request(request)
+        except oauth2.Error, e:
+            _logger.exception(u"Invalid LTI Request")
+            return HttpResponseBadRequest("Invalid LTI Request: " + e.message)
+
+        request.session['lti_parameters'] = provider.to_params()
```

Comment:
Storing this in the session with a fixed key means we can't have more than one question in the same unit, since they would override each others `lti_parameter`.  Also note that updates to the session are alway prone to race conditions, so we have to be very careful with concurrent requests.  I'm not sure yet what to do about it, since I don't understand all the details.  Maybe our easiest bet is to completely avoid redirecting, so we don't need the session.